tools/lsp/main.rs

```diff
@@ -187,77 +197,108 @@ fn main_loop(connection: &Connection, params: serde_json::Value) -> Result<(), E
         Some(if cli_args.style.is_empty() { "fluent".into() } else { cli_args.style });
     compiler_config.include_paths = cli_args.include_paths;
 
-    let mut document_cache = DocumentCache::new(compiler_config);
-    let request_queue = OutgoingRequestQueue::default();
+    let mut rh = RequestHandler::default();
+    register_request_handlers(&mut rh);
 
+    let request_queue = OutgoingRequestQueue::default();
     let server_notifier = ServerNotifier(connection.sender.clone(), request_queue.clone());
+    let ctx = Rc::new(Context {
+        document_cache: RefCell::new(DocumentCache::new(compiler_config)),
+        server_notifier: server_notifier.clone(),
+        init_param: init_params,
+    });
+
+    let mut futures = Vec::<Pin<Box<dyn Future<Output = Result<(), Error>>>>>::new();
+    let mut first_future = Box::pin(load_configuration(&ctx));
 
-    load_configuration(&server_notifier)?;
+    struct DummyWaker;
+    impl std::task::Wake for DummyWaker {
+        fn wake(self: Arc<Self>) {}
+    }
+    let waker = Arc::new(DummyWaker).into();
+    match first_future.as_mut().poll(&mut std::task::Context::from_waker(&waker)) {
+        Poll::Ready(x) => x?,
+        Poll::Pending => futures.push(first_future),
+    };
 
     for msg in &connection.receiver {
         match msg {
             Message::Request(req) => {
                 if connection.handle_shutdown(&req)? {
                     return Ok(());
                 }
-                handle_request(
-                    RequestHolder(req, server_notifier.clone()),
-                    &params,
-                    &mut document_cache,
-                )?;
+                futures.push(Box::pin(rh.handle_request(req, &ctx)));
             }
             Message::Response(resp) => {
-                let f = request_queue
-                    .lock()
-                    .unwrap()
-                    .remove(&resp.id)
-                    .ok_or("Response to unknown request")?;
-                f(resp, &mut document_cache);
+                if let Some(q) = request_queue.lock().unwrap().get_mut(&resp.id) {
+                    match q {
+                        OutgoingRequest::Done(_) => {
+                            return Err("Response to unknown request".into())
+                        }
+                        OutgoingRequest::Pending(x) => x.wake_by_ref(),
+                    };
+                    *q = OutgoingRequest::Done(resp)
+                } else {
+                    return Err("Response to unknown request".into());
+                }
             }
             Message::Notification(notification) => {
-                handle_notification(&server_notifier, notification, &mut document_cache)?
+                futures.push(Box::pin(handle_notification(notification, &ctx)))
             }
         }
+
+        let mut result = Ok(());
+        futures.retain_mut(|f| {
```

Comment:
Generally speaking, I wonder if this is the right approach. As long as the futures we generate resolve immediately on poll and all one by one then I guess it’s ok. But if any futures remain in the list, then the next time they’ll be polled is only when a new message is received. I wonder if this should either loop until all futures are resolved or use a runtime.

Reply:
Never mind, I forgot that the whole point of the exercise is to have futures outlive the current request. Still thinking polling futures and receiver at the same time might be cleaner in the long run. But I see that we don’t need that right now :). Sorry for the false alarm.

Reply:
Right now, all the future are pending on responses to requests, which happens only when we receive messages from the client and therefore get into this loop.

Reply:
Perhaps this deserves a comment